internal/server/v1/observations/derived_series.go

```diff
@@ -0,0 +1,507 @@
+//
+// Parse var name, which contains a variable and a set of filters.
+// For example: Person_Count[m=US_Census;p=P1Y].
+func parseVarName(varName string) (*varInfo, error) {
+	res := &varInfo{}
+
+	if strings.Contains(varName, "[") { // With filters.
+		if !strings.Contains(varName, "]") {
+			return nil, fmt.Errorf("missing ]")
+		}
+
+		leftBracketIndex := strings.Index(varName, "[")
+
+		res.statMetadata = &pb.StatMetadata{}
+		filterString := varName[leftBracketIndex+1 : len(varName)-1]
+		for _, filter := range strings.Split(filterString, ";") {
+			filterVal := filter[2:]
+			switch filter[0:1] {
+			case "m":
+				res.statMetadata.MeasurementMethod = filterVal
+			case "p":
+				res.statMetadata.ObservationPeriod = filterVal
+			case "u":
+				res.statMetadata.Unit = filterVal
+			case "s":
+				res.statMetadata.ScalingFactor = filterVal
+			}
+		}
+
+		res.statVar = varName[0:leftBracketIndex]
+
+	} else { // No filters.
+		res.statVar = varName
+	}
+
+	return res, nil
+}
+
+// Golang's AST package is used for parsing the formula, so we need to avoid sensitive tokens for
+// AST. For those tokens, we swap them with insensitive tokens before the parsing, then swap them
+// back after the parsing.
+var (
+	encodeForParseTokenMap = map[string]string{
+		"dc/":          "_DC_SLASH_",
+		"dcAggregate/": "_DC_AGGREGATE_SLASH_",
+		"[":            "_LEFT_SQUARE_BRACKET_",
+		"]":            "_RIGHT_SQUARE_BRACKET_",
+		"=":            "_EQUAL_TO_",
+		";":            "_SEMICOLON_",
+	}
+	decodeForParseTokenMap = map[string]string{
+		"_DC_SLASH_":             "dc/",
+		"_DC_AGGREGATE_SLASH_":   "dcAggregate/",
+		"_LEFT_SQUARE_BRACKET_":  "[",
+		"_RIGHT_SQUARE_BRACKET_": "]",
+		"_EQUAL_TO_":             "=",
+		"_SEMICOLON_":            ";",
+	}
+)
+
+func encodeForParse(s string) string {
+	res := s
+	for k, v := range encodeForParseTokenMap {
+		res = strings.ReplaceAll(res, k, v)
+	}
+	return res
+}
+
+func decodeForParse(s string) string {
+	res := s
+	for k, v := range decodeForParseTokenMap {
+		res = strings.ReplaceAll(res, k, v)
+	}
+	return res
+}
+
+func toVarCalculatorSeries(sourceSeries *pb.SourceSeries) *varCalculatorSeries {
+	series := &varCalculatorSeries{
+		statMetadata: &pb.StatMetadata{
+			MeasurementMethod: sourceSeries.GetMeasurementMethod(),
+			ObservationPeriod: sourceSeries.GetObservationPeriod(),
+			Unit:              sourceSeries.GetUnit(),
+			ScalingFactor:     sourceSeries.GetScalingFactor(),
+		},
+		points: []*pb.PointStat{},
+	}
+
+	var dates []string
+	for date := range sourceSeries.GetVal() {
+		dates = append(dates, date)
+	}
+	sort.Strings(dates)
+	for _, date := range dates {
+		series.points = append(series.points, &pb.PointStat{
+			Date:  date,
+			Value: sourceSeries.GetVal()[date],
+		})
+	}
+
+	return series
+}
+
+// Find common strings among a list of string sets.
+// For example, for input [{a,b,c},{a,c,d},{a,c,e}], it returns {a,c}.
+func commonStringsAmongStringSets(
```

Comment:
can use util.MergeDedupe()

Reply:
MergeDedupe() actually doesn't work here... 

For [{a,b,c},{a,c,d},{a,c,e}], MergeDedupe() will return {a,b,c,d,e}, but I want {a,c}.

Reply:
I see, then maybe move this function to util.go as stringSetIntersection()?

Reply:
done